test/golden_test.dart

```diff
@@ -53,20 +40,87 @@ void main() {
   });
 }
 
+
+extension on File {
+  String get fileName => path.basename(this.path);
+}
+
+extension on String {
+  List<String> get lines => this.trim().split('\n');
+}
+
+
+
+/// Returns the root of this project
+String get _projectRoot {
+  final scriptPath = Platform.script.toFilePath();
+  final scriptName = path.basename(scriptPath);
+  final scriptDirectory = path.join(scriptName, '..');
+  final projectRoot = path.normalize(scriptDirectory);
+  return projectRoot;
+}
+
+/// Returns the path to the Golden file of the given Test file.
+String _goldenFilePath(File testFile) {
+  final testFileName = testFile.fileName;
+  final goldenFileName = '$testFileName.golden';
+  final goldenPath = path.join(_goldensDirectory.path, goldenFileName);
+  return goldenPath;
+}
+
+/// Path to the grammar file
+final _grammarFilePath = path.join(_projectRoot, 'grammars', 'dart.json');
+
+
+
+/// Directory of Test Files
+final _testFileDirectory =
+    Directory(path.join(_projectRoot, 'test', 'test_files'));
+
+/// Directory of the Goldens
+final _goldensDirectory = Directory(path.join(_projectRoot, 'test', 'goldens'));
+
+/// The grammar file
+final _grammarFile = File(_grammarFilePath);
+
+/// Returns the test files within the test directory
+Iterable<File> get _testFiles => _testFileDirectory
+    .listSync()
+    .whereType<File>()
+    .where((file) => path.extension(file.path) == '.dart');
+
+/// Returns the Golden file of the Test file.
+File _getTheGoldenOfTestFile(File testFile) => File(_goldenFilePath(testFile));
+
+
+
+/// The Grammar
+Grammar _loadGrammar() {
+  final grammarString = _grammarFile.readAsStringSync();
+  final grammarWithFixedLookaheads = grammarString.replaceAll("(?>!", "(?!");
```

Comment:
I don't think we can make this kind of change here, there are other places parsing this grammar (such as in DevTools) that won't want to manipulate the grammar before parsing it. Is there a reason these changes can't be made directly in the grammar file instead?

Reply:
Ah that makes sense. But the trouble is: All editors have to use Oniguruma as it is the standard engine for the tmLanguage. (historically vscode didn't have a web version because they couldn't bring Oniguruma to web, once WASM got good enough it made it possible to move to web) and (also historically) the negative lookbehind is not a standard regular expression feature. So different engines have different grammars for it. Dart/JS engines use `(?!` grammar while Oniguruma uses `(?>!`. We either have to drop the use of the grammar altogether (which seems impossible in my view) or we have to convert the grammars for different engines. (And since this conversion is not that difficult, I say we have a loading point for the grammar that does the job and use it everywhere)

Reply:
There are also two other options, To use Oniguruma directly without bridging it. (at end, the code is going to run on it) so the vscode team has two packages:

- One is [vscode-oniguruma](https://github.com/microsoft/vscode-oniguruma) that directly binds the C++ version to the Electron/Node

- The other is [wasm-oniguruma](https://github.com/rebornix/wasm-oniguruma) which is used for runnig on the web. 

Maybe the DevTools can harness these ?

Reply:
> Aalso historically; the negative lookbehind is not a standard regular expression feature. So different engines have different grammars for it. Dart/JS engines use `(?!` while Oniguruma has `(?>!`. And so it makes it impossible to directly load the grammars from one side to the other.

Thanks for the info - I'm trying to read up on this to think about the best option - although I can't find `(?>!` listed in https://github.com/kkos/oniguruma/blob/master/doc/RE (I do see both `(?!subexp)` and ` (?<!subexp)`). Am I looking in the right place?